### ISSUE TITLE
Question on allowing to run hyrise in a non global/singleton mode

### ISSUE THREAD

**Issue by jackvincenow** [hyrise/hyrise#2289]
Hello Hyrise Team.

At the moment, there exists a single, global, Hyrise object. I wish to instead parametrize it, so that whenever somebody tries to access it, instead of accessing the global singleton, he will instead access a Hyrise typed object which passed along as either a class member or a function parameter. After some fiddling with the code, I believe that this will take about 10 days to implement, after which we can go over the PR together in case there's anything you want to change. The benefits of this are that it allows running Hyrise in parallel on several different storages, as well as allows an easy implementation of separate schemas.

Will love to get your feedback, and a green light for me to proceed.
have a great day.

**Comment by mrks** [comment]
I have a couple of quick thoughts and questions about this.

* Currently, the Hyrise singleton holds both truly global state (e.g., the scheduler, loaded plugins) as well as instance-specific state (most importantly, the storage manager). To allow for multiple instances, those would have to be unwoven.
* The reason we have that singleton is because it allows for easy access to that state. In a previous database that I worked on, that state had to be passed around the entire code base, which did not do the interfaces a favor. Optimally, we could avoid that.
* Who do you envision to hold the different Hyrise objects? I could see a `USE (schema_name)` command that selects a schema (and thus a Hyrise object) for a given connection. In that case, the connection would hold the instances and would pass the selected one into the SQLPipeline.
* In your use case, how would the Hyrise instances differ? You mention "several different storages", could you describe what you are trying to achieve? This sounds somewhat familiar to something that we work on internally, maybe we can identify some overlap.

**Comment by jackvincenow** [comment]
1) Agreed. Since most of the Hyrise code is in classes, this can be handle as a member. In our internal tests, we saw that putting a pointer to StorageManager in StoredTableNode and in AbstractRule would avoid most cases of having to add an external parameter to functions.

2) In our experimental implementation we passed a parameter to the SQLPipeline, we did not support any use case involving a command to select a schema. I would guess that USE makes sense here.

3) We have three cases: 1) We want a different system to hold schema data instead of Hyrise, and for Hyrise to be able to access that system via an adapter object that implements a StorageManagerInterface . 2) We want to use the same Hyrise instance to host several databases without them seeing each other. 3) We have our own transaction system and we want to expose to Hyrise the schema matching the current transaction.
